Use AuthenticationTokens

vault opened the pull request:
Opening for initial feedback.

This is based on #92 and moves to using the AuthenticationTokens plugin as was suggested in a couple of other pull-requests (#69, #68, and #94). This also includes support for client certificates like #94.

casz commented:
@vault think you need to rebase onto PR #92 again and resolve conflicts

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketClientCertificateAuthenticator.java:
```diff
@@ -0,0 +1,35 @@
+        try {
+            builder.setSSLContext(buildSSLContext());
+        } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException | KeyManagementException ignored) {
+
```
I think we'll need to either pass a  `TaskListener` to these methods or log the exceptions. Certainly unacceptable to ignore and continue.

May even need to wrap as an `IOException` and throw

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,22 @@
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.client.HttpClientBuilder;
+
+public abstract class BitbucketAuthenticator<T extends Credentials> {
```
On the plus side, you are implementing a strategy

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,22 @@
+
+public abstract class BitbucketAuthenticator<T extends Credentials> {
+
+    public final T credentials;
```
On the minus side you are basically wrapping a Credentials... please don't do that, extract what you need from the Credential and move on, that way the `BitbucketAuthenticator` can be used from remote agents rather than only from the master JVM

stephenc replied:
Yes that means you need a subclass for each type of credential that is supported... but you needed that anyway. Also... just say no to public fields

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:
```diff
@@ -64,71 +97,63 @@
 
 public class BitbucketCloudApiClient implements BitbucketApi {
     private static final Logger LOGGER = Logger.getLogger(BitbucketCloudApiClient.class.getName());
+    private static final HttpHost API_HOST = HttpHost.create("https://api.bitbucket.org");
```
I strongly recommend waiting until after #80 is merged (or else rebasing this PR on top of #80)

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpoint.java:
```diff
@@ -112,15 +115,15 @@ public final String getCredentialsId() {
+                        StandardCredentials.class,
                         Jenkins.getActiveInstance(),
                         ACL.SYSTEM,
                         URIRequirementBuilder.fromUri(getServerUrl()).build()
```
You will also need to use the `AuthenticationTokens` matcher. (Further, it may be better to use the "context" variant... so we can decorate with the URL scheme and restrict which Authentication token types are valid against different URLs... i.e. you cannot have client certificates for `http` but only for `https` but that can be handled later as a finesse)

casz replied:
CredentialsMatchers are used below so all looks good

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/AbstractBitbucketEndpointDescriptor.java:
```diff
@@ -57,9 +58,12 @@ public ListBoxModel doFillCredentialsIdItems(@QueryParameter String serverUrl) {
+                StandardCredentials.class,
                 URIRequirementBuilder.fromUri(serverUrl).build(),
-                CredentialsMatchers.instanceOf(StandardUsernamePasswordCredentials.class)
+                CredentialsMatchers.anyOf(
```
No, you should be using `AuthenticationTokens.matcher(BitbucketAuthenticator.class)` here. No need to test for `instanceOf` any more. In fact this switch is the key reason to move to AuthenticationTokens... as it means an extension plugin can register more authenticator sources and then the support wires right through

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/endpoints/BitbucketCloudEndpoint.java:
```diff
@@ -23,14 +23,15 @@
-import javax.annotation.Nonnull;
 import org.kohsuke.stapler.DataBoundConstructor;
 
+import javax.annotation.Nonnull;
```
There is no published JSR 305. Please use `edu.umd.cs.findbugs.annotations.NonNull`

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java:
```diff
@@ -45,12 +46,43 @@
+import org.apache.http.message.BasicNameValuePair;
+import org.apache.http.util.EntityUtils;
+import org.codehaus.jackson.type.TypeReference;
+
```
all imports in one section sorted alphabetically: https://github.com/jenkinsci/scm-api-plugin/blob/master/CONTRIBUTING.md details the code style

stephenc commented on src/test/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketMockApiFactory.java:
```diff
@@ -2,15 +2,16 @@
 import hudson.Extension;
 import hudson.ExtensionList;
+import org.apache.commons.lang.StringUtils;
+
```
Check your IDE settings for import sorting... it's all of them not just the two I have tagged

casz commented:
@vault this should now be in a good state to rebase 👍

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java:
```diff
@@ -636,9 +651,11 @@ public ListBoxModel doFillCredentialsIdItems(@AncestorInPath SCMSourceOwner cont
+                    StandardCredentials.class,
                     URIRequirementBuilder.fromUri(bitbucketServerUrl).build(),
-                    CredentialsMatchers.anyOf(CredentialsMatchers.instanceOf(StandardUsernamePasswordCredentials.class))
+                    CredentialsMatchers.anyOf(
```
`AuthenticationTokens.matcher(BitbucketCredentials.class)` should suffice instead of the credentials matchers

NOTE: you will need to use an `AuthenticationTokenContext` to handle the different types of auth for `http` vs `https` urls, so at that point you will change to `AuthenticationTokens.matcher(BitbucketCredentials.class, authContext)`

NOTE2: This comment applies to everywhere you are using `CredentialsMatchers.instanceOf`

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java:
```diff
@@ -1125,9 +1133,17 @@ public String getDisplayName() {
+                        CredentialsProvider.lookupCredentials(StandardCertificateCredentials.class,
+                                context,
+                                context instanceof Queue.Task ? Tasks.getDefaultAuthenticationOf((Queue.Task) context) : ACL.SYSTEM,
+                                URIRequirementBuilder.fromUri(bitbucketServerUrl).build()), CredentialsMatchers.withId(value)) != null) {
```
You will need both `withId` and `AuthenticationTokens.matcher(BitbucketCredentials.class)` to handle the case where there are two different types of credentials with the same id at different layers in the folder hierarchy.

stephenc replied:
The point being that the drop-down list will be showing only those that `AuthenticationTokens.matcher(BitbucketCredentials.class)` so may hide higher up ones that happen to have the same id but don't match the type we want...

say at the root there is a userpass with id `foo` and one folder down there is a certificate with id `foo`... if the url is `http` we should be able to select `foo` and mean the userpass...

and such a config could actually be desired, e.g. against `https` we want to use the stronger client cert... against `http` we have no choice... and we set up everything to always have id of `foo` so that the choice is easy... not a config I'd recommend... but it is logical... and i think credentials-api would support it

vault commented:
I think I've addressed most of the requests here. The main omissions being I'm not sure what else to do in the client-cert exception case other than log something. What other changes do you think this needs? Any thoughts on how to handle that case correctly? I looked into `TaskListener` a bit but I wasn't able to figure out the right way to use it here.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public static final String PROTOCOL_HTTPS = "HTTPS";
+
+
```
:ghost: whitespace

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public void configureBuilder(HttpClientBuilder builder) { }
+
+
```
:ghost: whitespace

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public void configureContext(HttpClientContext context, HttpHost host) { }
+
+
```
:ghost: whitespace

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,105 @@
+     */
+    public void configureRequest(HttpRequest request) { }
+
+
```
:ghost: whitespace

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * The URL protocol as reported in an {@link AuthenticationTokenContext}
+     */
+    public static final String PROTOCOL_PURPOSE = "PROTOCOL";
```
I would call this `Scheme`

vault replied:
For the constant name, or its string value? Also, `<blah>_PURPOSE` reads kind of weird. Is there a naming scheme that would be more preferred than that?

stephenc replied:
I’d use `SCHEME = “scheme”` or `SERVER_URI_SCHEME = “bitbucket.server.uri.scheme”`

The length of the constant doesn’t use up more space as it will still be an 8 byte reference to the constant pool anyway, so more descriptive is better

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * The purpose value for HTTP instances
+     */
+    public static final String PROTOCOL_HTTP = "HTTP";
```
just use the URI scheme rather than a constant (i.e. parse the URL, or just substring up to the first `:` and lowercase the scheme part)... if you want to have constants for HTTP and HTTPS their values should be the same as an URL (i.e. lowercase)

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * The Bitbucket instance type as reported in an {@link AuthenticationTokenContext}
+     */
+    public static final String INSTANCE_TYPE_PURPOSE = "INSTANCE_TYPE";
```
Make it clear that this is bitbucket, a better value would be `Bitbucket.Type`

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+     * @param serverUrl The URL being authenticated against
+     * @return an {@link AuthenticationTokenContext} for use with the AuthenticationTokens APIs
+     */
+    public static AuthenticationTokenContext<BitbucketAuthenticator> authenticationContext(String serverUrl) {
```
Nice

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+        boolean isCloud = serverUrl == null || serverUrl.equals(BitbucketCloudEndpoint.SERVER_URL);
+
+        return AuthenticationTokenContext.builder(BitbucketAuthenticator.class)
+                .with(PROTOCOL_PURPOSE, isHttps ? PROTOCOL_HTTPS : PROTOCOL_HTTP)
```
I think you could be perfectly fine not to parse the URL and just put the URL in the token context as `url` and be done... let the impls decide if the url is cloud or server...

fine to do it here, but I would recommend adding the URL to the token context also in that case, as you don't know what will need it further down the road

(for example a credential may only be valid against a specific server)

rsandell commented:
I'm not sure what @stephenc 's philosophy has been on this, but there are signature changes in a couple of unrestricted public classes and extension points _(constructor changes, methods removed etc.)_
In general we make an effort to keep old signatures _(but deprecated)_ and make them work with the new code.

vault commented:
I should be able to fix that. I'll try to annotate changes that are problematic and try to get them fixed in the next day or so.

vault commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketApiFactory.java:
```diff
@@ -44,20 +43,20 @@ protected abstract BitbucketApi create(@Nullable String serverUrl,
     @NonNull
     public static BitbucketApi newInstance(@Nullable String serverUrl,
-                                           @Nullable StandardUsernamePasswordCredentials credentials,
+                                           @Nullable BitbucketAuthenticator authenticator,
```
Changed public method signature. Should be fixable by having a method with the old signature that calls this with an authenticator instance created from the passed credentials.

casz replied:
Someone might have written an extension, please consider adding the ability to accept StandardUsernamePasswordCredentials and convert them to BitbucketAuthenticator

casz replied:
nvm already done

vault commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketApiFactory.java:
```diff
@@ -28,14 +27,14 @@
     @NonNull
     protected abstract BitbucketApi create(@Nullable String serverUrl,
-                                           @Nullable StandardUsernamePasswordCredentials credentials,
+                                           @Nullable BitbucketAuthenticator authenticator,
```
This is a protected method with a changed signature. Is that a concern for API compatibility? In theory there could be other sub-classes outside the ones this plugin provides.

casz replied:
Someone might have written an extension, please consider adding the ability to accept StandardUsernamePasswordCredentials and convert them to BitbucketAuthenticator

casz replied:
Nvm already done 😇

vault commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:
```diff
@@ -132,12 +131,8 @@
-        } else {
-            this.credentials = null;
-        }
+    public BitbucketCloudApiClient(String owner, String repositoryName, BitbucketAuthenticator authenticator) {
```
Changed public method signature.

vault replied:
Included a deprecated compatible implementation.

vault commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:
```diff
@@ -223,13 +215,6 @@ public String getRepositoryUri(@NonNull BitbucketRepositoryType type,
     }
 
-    @CheckForNull
-    public String getLogin() {
```
Removed public method.

Adding it back in likely won't have the exact same semantics since I think we'd only be able to return the credential ID from the authenticator instead of the actual username. Would that still be acceptable?

casz replied:
Seems you provided a deprecated method, resolving conversation

vault commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java:
```diff
@@ -148,16 +148,15 @@
-                                    @CheckForNull StandardUsernamePasswordCredentials creds, boolean userCentric) {
-        this.credentials = (creds != null) ? new UsernamePasswordCredentials(creds.getUsername(),
-                Secret.toString(creds.getPassword())) : null;
+                                    @CheckForNull BitbucketAuthenticator authenticator, boolean userCentric) {
```
Changed public method signature.

vault replied:
Included a compatible deprecated implementation.

vault commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMNavigator.java:
```diff
@@ -609,8 +616,21 @@ public ListBoxModel doFillServerUrlItems() {
         }
 
-        public FormValidation doCheckCredentialsId(@QueryParameter String value) {
+        public FormValidation doCheckCredentialsId(@AncestorInPath SCMSourceOwner context, @QueryParameter String serverUrl, @QueryParameter String value) {
```
Are DescriptorImpl classes considered part of the public API that this plugin is exposing?

casz replied:
no

vault commented:
Rebased and added compatibility methods for what I think are all the important places this branch wasn't backwards compatible.

Let me know if it seems like there's something I missed.

michelzanini commented:
Apart from username/password and client certificates, does this PR also supports using a OAuth token to connect to the REST API of Bitbucket?

vault commented:
This PR doesn't add support for OAuth tokens, but it should make adding that support easier.

andrewspiers commented:
Is the only thing blocking this @stephenc 's review of 4th January?

casz commented:
@vault please rebase 😇

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api;
```
License

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketClientCertificateAuthenticator.java:
```diff
@@ -0,0 +1,56 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api.credentials;
```
License

vault replied:
Added.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/BitbucketAuthenticator.java:
```diff
@@ -0,0 +1,101 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api;
```
License

vault replied:
Added.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketUsernamePasswordAuthenticator.java:
```diff
@@ -0,0 +1,48 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api.credentials;
```
License

vault replied:
Added.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketClientCertificateAuthenticatorSource.java:
```diff
@@ -0,0 +1,46 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api.credentials;
```
License

vault replied:
Added.

casz commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketUsernamePasswordAuthenticatorSource.java:
```diff
@@ -0,0 +1,31 @@
+package com.cloudbees.jenkins.plugins.bitbucket.api.credentials;
```
License

vault replied:
Added.

stephenc commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java:
```diff
@@ -614,18 +617,25 @@ public BitbucketTeam call() throws Exception {
+        cacheKey.append(owner);
+
+        if (authenticator != null) {
+            cacheKey.append("::").append(authenticator.getId());
```
So.... the change from username to credentials id may mean that we lose the cache where there are two credentials instances with the same username... OTOH that might have been a way to capture information if we had been caching more than repositories... so likely not a bad thing... and this cache is internal, so if needs be could be refactored to a different key later.

casz commented:
Just to cover @hoegertn comment about being unable to use for Bitbucket Cloud is not true.
Create an atlassian user account using username & password credentials with username being the email of the user seems to be working just fine.

vault commented:
Rebased and added license headers which I think addresses all outstanding issues.
